src/common/dav/tx.c

```diff
@@ -398,35 +364,144 @@ tx_alloc_common(struct tx *tx, size_t size, type_num_t type_num,
 		goto err_oom;
 
-	ret = action->heap.offset;
-	size = action->heap.usable_size;
-
-	const struct tx_range_def r = {ret, size, args.flags};
-
-	if (tx_ranges_insert_def(pop, tx, &r) != 0)
+	palloc_get_prange(action, &off, &size, 1);
+	r = &(struct tx_range_def){off, size, args.flags};
+	if (tx_ranges_insert_def(pop, tx, r) != 0)
 		goto err_oom;
 
-	return ret;
+	return action->heap.offset;
 
 err_oom:
 	tx_action_remove(tx);
 	D_CRIT("out of memory\n");
 	return obj_tx_fail_null(ENOMEM, args.flags);
 }
 
+/*
+ * tx_create_wal_entry -- convert to WAL a single ulog UNDO entry
+ */
+int
+tx_create_wal_entry(struct ulog_entry_base *e, void *arg,
+		    const struct mo_ops *p_ops)
+{
+	/* suppress unused-parameter errors */
+	SUPPRESS_UNUSED(arg);
+
+	int			 rc = 0;
+	uint64_t		 offset = ulog_entry_offset(e);
+	daos_size_t		 dst_size = sizeof(uint64_t);
+	struct ulog_entry_val	*ev;
+	struct ulog_entry_buf	*eb;
+	uint64_t		 v;
+	uint64_t		*dst;
+
+	D_ASSERT(p_ops->base != NULL);
+	dst = (uint64_t *)((uintptr_t)((dav_obj_t *)p_ops->base)->do_base + offset);
+
+	switch (ulog_entry_type(e)) {
+#ifdef	WAL_SUPPORTS_AND_OR_OPS
+	case ULOG_OPERATION_AND:
+		ev = (struct ulog_entry_val *)e;
+		v = ev->value;
+
+		rc = dav_wal_tx_and(p_ops->base, dst, v);
+		break;
+	case ULOG_OPERATION_OR:
+		ev = (struct ulog_entry_val *)e;
+		v = ev->value;
+
+		rc = dav_wal_tx_or(p_ops->base, dst, v);
+		break;
+#else
+	case ULOG_OPERATION_CLR_BITS:
+		ev = (struct ulog_entry_val *)e;
+		v = ev->value;
+
+		rc = dav_wal_tx_clr_bits(p_ops->base, dst, ULOG_ENTRY_VAL_TO_POS(v),
+					 ULOG_ENTRY_VAL_TO_BITS(v));
+		break;
+	case ULOG_OPERATION_SET_BITS:
+		ev = (struct ulog_entry_val *)e;
+		v = ev->value;
+
+		rc = dav_wal_tx_set_bits(p_ops->base, dst, ULOG_ENTRY_VAL_TO_POS(v),
+					 ULOG_ENTRY_VAL_TO_BITS(v));
+		break;
+#endif
+	case ULOG_OPERATION_SET:
+		ev = (struct ulog_entry_val *)e;
+
+		rc = dav_wal_tx_snap(p_ops->base, dst, dst_size, (void *)&ev->value, 0);
+		break;
+	case ULOG_OPERATION_BUF_SET:
+		eb = (struct ulog_entry_buf *)e;
+
+		dst_size = eb->size;
+		rc = dav_wal_tx_set(p_ops->base, dst, 0, dst_size);
+		break;
+	case ULOG_OPERATION_BUF_CPY:
+		eb = (struct ulog_entry_buf *)e;
+
+		dst_size = eb->size;
+		/* The only undo entry from dav that needs to be
+		 * transformed into redo
+		 */
+		rc = dav_wal_tx_snap(p_ops->base, dst, dst_size, dst, 0);
+		break;
+	default:
+		ASSERT(0);
+	}
+
+	return rc;
+}
+
+int
+lw_tx_begin(dav_obj_t *pop)
+{
+	struct umem_wal_tx	*utx = pop->do_utx;
+	int			 rc;
+
+	D_ASSERT(utx == NULL);
+	utx = dav_umem_wtx_new(pop);
+	if (utx == NULL)
+		return obj_tx_fail_err(EINVAL, 0);
+	rc = dav_wal_tx_reserve(pop);
+	if (rc) {
+		D_ERROR("so_wal_reserv failed, "DF_RC"\n", DP_RC(rc));
+		D_FREE(utx);
+		pop->do_utx = NULL;
+	}
+	return rc;
+}
+
+int
+lw_tx_end(dav_obj_t *pop)
+{
+	struct umem_wal_tx	*utx;
+	int			 rc;
+
+	utx = pop->do_utx;
+	D_ASSERT(utx != NULL);
+
+	rc = dav_wal_tx_commit(pop);
```

Comment:
wal_commit() will yield, so if other ULT start transaction while this one yielding, the other ULT will use the pop->do_utx?

Reply:
Right, this is not ULT safe and must be fixed.